### ISSUE TITLE
Module shared_vpc_access causes v >= 12.0 causes unnecessary changes

### ISSUE THREAD

**Issue by cpf-hse** [terraform-google-modules/terraform-google-project-factory#750]
### TL;DR

When switching from Version 11.3.1 to 13.0 terraform plan delivered 34 to add and 34 to destroy without any other changes.
Going back to version 11.3.1 terraform plan shows "No Changes".

We also upgraded the provider version from 3.5.8 to >= 4.0.0

All changes looked like this
(Please not I renamed real project names, project ids, subnets etc.)

module.shared_vpc.module.shared_vpc_access["PROJECT1"].google_compute_subnetwork_iam_member.service_shared_vpc_subnet_users[0] must be replaced
-/+ resource "google_compute_subnetwork_iam_member" "service_shared_vpc_subnet_users" {
~ etag = "BwW9kXb6SBI=" -> (known after apply)
~ id = "projects/our-shared-network/regions/europe-west3/subnetworks/subnet-1/roles/compute.networkUser/serviceaccount:[service-PROJECTID@container-engine-robot.iam.gserviceaccount.com](mailto:service-PROJECTID@container-engine-robot.iam.gserviceaccount.com)" -> (known after apply)
~ subnetwork = "projects/our-shared-network/regions/europe-west3/subnetworks/subnet1" -> "subnet1" # forces replacement
# (4 unchanged attributes hidden)
}

### Expected behavior

No changes. Your infrastructure matches the configuration.

### Observed behavior

Plan: 34 to add, 0 to change, 34 to destroy.

### Terraform Configuration

```hcl
module "shared_vpc_access" {
  for_each = var.service_projects
  source   = "terraform-google-modules/project-factory/google//modules/shared_vpc_access"
  version  = ">= 12.0.0"

  enable_shared_vpc_service_project = true
  host_project_id                   = var.host_project
  service_project_id                = each.key
  active_apis = [
    "compute.googleapis.com",
    "container.googleapis.com"
  ]
  shared_vpc_subnets = [
    "projects/our-shared-network/regions/europe-west3/subnetworks/subnet-1",
    "projects/our-shared-network/regions/europe-west3/subnetworks/subnet-2",
  ]

  depends_on = [
    google_project_service.host_vpc_container,
    google_project_service.example-host_vpc_servicenetworking
  ]
}
```

### Terraform Version

```sh
terraform {
  required_version = ">= 1.0"
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = ">= 4.0.0"
    }
    google-beta = {
      source  = "hashicorp/google-beta"
      version = ">= 4.0.0"
    }
  }
  backend "http" {
  }
}

provider "google" {
  region = var.gcp_region
}

provider "google-beta" {
  region = var.gcp_region
}
```

**Comment by saez0pub** [comment]
Hello,

The commit with changed this behaviour is https://github.com/terraform-google-modules/terraform-google-project-factory/commit/71ea320dca2e6b56c800cd8f55e90c512224933d#diff-bbb693355b1cd271114cf1dba901038f82a148de82ce55ef93bd73c4f37dde4cR53

On this line: https://github.com/terraform-google-modules/terraform-google-project-factory/blob/71ea320dca2e6b56c800cd8f55e90c512224933d/modules/shared_vpc_access/main.tf#L53

We should have a backward compatible option for this kind of change. If I want to upgrade to the latest version, I need to recreate my kube cluster as a side effect of this change 😭 .

**Comment by saez0pub** [comment]
Forget it, that's not the commit I mentioned, I don't know why the split was not done before.
Would you recommend me to edit the state or rm then import ?

**Comment by saez0pub** [comment]
It seems that the order of subnetwork_api has changed 😢 so count.index does not match the the previous entries.

**Comment by cpf-hse** [comment]
I am glad I am not the only one with this issue.

But I am not confidend that I keep relying on / using google terraform modules. Seems like no one cares to fix this issue.

**Comment by saez0pub** [comment]
Hello,

This issue is not simple.
I think that the main probleme is the usage of count.index iterating on a list instead of using foreach. The index Can change depending to the user input.
So another breaking change is needed to fix totally this issue 

In ordre to get out of the problem, I did a terraform state rm then plan and import and it's ok.

**Comment by kristinafl** [comment]
@aaron-lane can you advise on this issue please? @edwardmedia ((https://github.com/edwardmedia) suggested that this issue relates to the module and the module owners can assist - see here: https://github.com/hashicorp/terraform-provider-google/issues/12759. Thanks!
